### ISSUE TITLE
Tensorflow Program Hangs on pthread_cond_wait

### ISSUE THREAD

**Issue by benbotto** [tensorflow/tensorflow#20346]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04.4 LTS, xenial
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**:  tensorflow-gpu (1.8.0)
- **Python version**: Python 3.5.2
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 9.0.176/7.1.3.16
- **GPU model and memory**: GeForce GTX 850M with 2GB of memory
- **Exact command to reproduce**:

### Describe the problem
I have a program using Tensorflow and it consistently hangs after a week or so, arbitrarily.  The GPU utilization goes to 0%, and the program stops training.  A stack trace shows that the program is stuck waiting on pthread_cond_wait.

### Source code / logs
I don't know that the source code will be of much use since it's not a concise example that reproduces the problem, but the code is available here: https://github.com/benbotto/bsy-dqn-atari/tree/breakout-best  My code is single threaded.

When last the program hung I took a stack trace.  That's available here: https://pastebin.com/LiPhz2CE

This looks similar to this report: https://github.com/tensorflow/tensorflow/issues/1947
 
Let me know if more information is needed.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Jackhuang945** [comment]
the same problem, no solution now.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rohan100jain** [comment]
Looked at the pastebin and nothing obvious jumped out to me. @mrry any ideas?

If the problem happened again, could you paste a newer version of the stack trace?

**Comment by benbotto** [comment]
The hang happens reliably on one of my machines.  I have another machine that's running the same software on the same version of Tensorflow--albeit different hardware--and I haven't had any hangs on that one.  The hang is sporadic, though, and often takes a week or more.  If I really need to I can take another stack trace, but that will take some time.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @rohan100jain, @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mrry** [comment]
Looking at this again, and there's no obvious answer in the stack trace, but I have a few suggestions:

1.  Can you try creating your `tf.Session` with `config=tf.ConfigProto(inter_op_parallelism_threads=1, intra_op_parallelism_threads=1)`? This shouldn't fix anything... if anything it makes it more likely that buggy code will deadlock sooner... but it should give a more compact thread dump to analyze.
2. Are you using `tf.py_func()` in your code? It looks like "Thread 1" (which appears twice in the pastebin link) is blocked in some NumPy code, in what appears to be OpenBLAS (`exec_blas_async_wait()`). It's possible that if your other machine has a different version of NumPy, then it might have different behavior for this method.
3. Related to (2), experimenting with the `OMP_NUM_THREADS` variable (e.g. setting it to 1?) might reveal something about the behavior.

If you can reproduce the hang and capture another (ideally smaller) stack trace, that would be useful!

**Comment by mrry** [comment]
It could be related to this issue in OpenBLAS: https://github.com/xianyi/OpenBLAS/issues/660

[This comment](https://github.com/xianyi/OpenBLAS/issues/660#issuecomment-277033954) suggests that setting the environment variable `OPENBLAS_NUM_THREADS=1` is a possible workaround, but may slow things down.

**Comment by benbotto** [comment]
Thanks for the response, @mrry.

1. Right now I'm not actually creating a session explicitly as I'm using tf.keras which evidently creates a session behind the scenes.  It looks like Keras has the option of using a manually defined session (`tf.keras.backend.set_session`), so I'll try your config recommendation and see if I can get a better stack dump.
2. I'm not using `tf.py_func` anywhere, no.  As far as I know both machines are using identical version of everything (Tensorflow, Keras, numpy, and all other dependencies).  I'm using virtualenv, and I set up both machines simultaneously.  I'll double check the np version, though.

I wish I knew how to set up a smaller test case that would reproduce the issue--as a programmer I certainly understand how important that is--but I'm not sure how in this instance.  I don't think my code is doing anything out of the ordinary: I'm generating images, doing some basic processing on those (crop and grayscale and such), then feeding them through a rather trivial network.  I'm not _directly_ using any threading in my code or anything else that would cause a deadlock.  All that is to say I wish I could be of more help (and furthermore I wish the code would run reliably on all my machines!).

Anyway, I'll try again with the session configured as recommended and see if I can get another stack trace.  It may take a few weeks to lock up again.

**Comment by mrry** [comment]
@formath That sounds like it could be a different problem. However, there isn't enough information in that one stack to diagnose the cause: it just tells us that one call to `sess.run()` is hanging somewhere. Can you please open a new issue with a minimal reproducible example and/or a full set of thread stacks? Also, to remove a potential source of complexity, please try to reproduce it using `tf.Session()` instead of `tf.Session(rpc_server)` (although if it only reproduces in an RPC-based session, then that would tell us something useful).

**Comment by tensorflowbutler** [comment]
Nagging Assignees @rohan100jain, @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 44 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
We are closing this issue for now due to lack of activity. Please comment if this is still an issue for you. Thanks!
